pkg/serviceaccount/claims_test.go

```diff
@@ -188,6 +205,154 @@ func TestClaims(t *testing.T) {
 				},
 			},
 		},
+		{
+			// node with feature gate disabled
+			sa:   sa,
+			node: node,
+			// really fast
+			exp: 0,
+			// nil audience
+			aud: nil,
+
+			sc: &jwt.Claims{
+				Subject:   "system:serviceaccount:myns:mysvcacct",
+				IssuedAt:  jwt.NewNumericDate(time.Unix(1514764800, 0)),
+				NotBefore: jwt.NewNumericDate(time.Unix(1514764800, 0)),
+				Expiry:    jwt.NewNumericDate(time.Unix(1514764800, 0)),
+			},
+			pc: &privateClaims{
+				Kubernetes: kubernetes{
+					Namespace: "myns",
+					Svcacct:   ref{Name: "mysvcacct", UID: "mysvcacct-uid"},
+				},
+			},
+		},
+		{
+			// node & pod with feature gate disabled
+			sa:   sa,
+			node: node,
+			pod:  pod,
+			// really fast
+			exp: 0,
+			// nil audience
+			aud: nil,
+
+			sc: &jwt.Claims{
+				Subject:   "system:serviceaccount:myns:mysvcacct",
+				IssuedAt:  jwt.NewNumericDate(time.Unix(1514764800, 0)),
+				NotBefore: jwt.NewNumericDate(time.Unix(1514764800, 0)),
+				Expiry:    jwt.NewNumericDate(time.Unix(1514764800, 0)),
+			},
+			pc: &privateClaims{
+				Kubernetes: kubernetes{
+					Namespace: "myns",
+					Pod:       &ref{Name: "mypod", UID: "mypod-uid"},
+					Svcacct:   ref{Name: "mysvcacct", UID: "mysvcacct-uid"},
+				},
+			},
+		},
+		{
+			// node alone
+			sa:   sa,
+			node: node,
+			// enable node binding feature
+			featureNodeBinding: true,
+			// really fast
+			exp: 0,
+			// nil audience
+			aud: nil,
+
+			sc: &jwt.Claims{
+				Subject:   "system:serviceaccount:myns:mysvcacct",
+				IssuedAt:  jwt.NewNumericDate(time.Unix(1514764800, 0)),
+				NotBefore: jwt.NewNumericDate(time.Unix(1514764800, 0)),
+				Expiry:    jwt.NewNumericDate(time.Unix(1514764800, 0)),
+			},
+			pc: &privateClaims{
+				Kubernetes: kubernetes{
+					Namespace: "myns",
+					Svcacct:   ref{Name: "mysvcacct", UID: "mysvcacct-uid"},
+					Node:      &ref{Name: "mynode", UID: "mynode-uid"},
+				},
+			},
+		},
+		{
+			// node and pod
+			sa:   sa,
+			pod:  pod,
+			node: node,
+			// enable embedding pod node info feature
+			featurePodNodeInfo: true,
+			// really fast
+			exp: 0,
+			// nil audience
+			aud: nil,
+
+			sc: &jwt.Claims{
+				Subject:   "system:serviceaccount:myns:mysvcacct",
+				IssuedAt:  jwt.NewNumericDate(time.Unix(1514764800, 0)),
+				NotBefore: jwt.NewNumericDate(time.Unix(1514764800, 0)),
+				Expiry:    jwt.NewNumericDate(time.Unix(1514764800, 0)),
+			},
+			pc: &privateClaims{
+				Kubernetes: kubernetes{
+					Namespace: "myns",
+					Svcacct:   ref{Name: "mysvcacct", UID: "mysvcacct-uid"},
+					Pod:       &ref{Name: "mypod", UID: "mypod-uid"},
+					Node:      &ref{Name: "mynode", UID: "mynode-uid"},
+				},
+			},
+		},
+		{
+			// node and secret (secret takes precedence)
```

Comment:
🤔 this seems weird to me.  Shouldn't we error?

Reply:
this function cannot error at the moment - we have a `switch` statement in the tokens `Create` handler which will ensure we never actually pass both instead.

Do we want to change the function signature here and make the `Claims` function more defensive?

Reply:
I think `panic`ing on invariant violation is okay here.

Reply:
Granted on this case, we might be able to use generics to represent the input constraints?

Reply:
Worked on this a bit, but due to the inability to specify generic 'vars', this code gets quite messy (especially when it comes to refactoring our table based tests).

Let's follow up with this, or at least get a PoC of it together in time for beta/GA so we have more confidence in how this function is used.

Reply:
done